File: docs/_code/energy_topic.cpp
#include <diaspora/Driver.hpp>
#include <diaspora/Metadata.hpp>
#include <diaspora/TopicHandle.hpp>
#include <mofka/MofkaDriver.hpp>
#include <iostream>

int main(int argc, char** argv) {

    if(argc != 2) {
        std::cerr << "Usage: "
                  << argv[0] << " <groupfile>" << std::endl;
        return -1;
    }

    auto group_file = argv[1];

    try {

        diaspora::Metadata options;
        options.json()["group_file"] = group_file;
        options.json()["margo"] = nlohmann::json::object();
        options.json()["margo"]["use_progress_thread"] = true;

        diaspora::Driver driver = diaspora::Driver::New("mofka", options);

        // START CREATE TOPIC
        diaspora::Validator validator =
            diaspora::Validator::FromMetadata(
                diaspora::Metadata{{
                    {"type","energy_validator:libenergy_validator.so"},
                    {"energy_max", 100}}}
            );

        diaspora::PartitionSelector selector =
            diaspora::PartitionSelector::FromMetadata(
                diaspora::Metadata{{
                    {"type","energy_partition_selector:libenergy_partition_selector.so"},
                    {"energy_max", 100}}}
            );

        diaspora::Serializer serializer =
            diaspora::Serializer::FromMetadata(
                diaspora::Metadata{{
                    {"type","energy_serializer:libenergy_serializer.so"},
                    {"energy_max", 100}}}
            );

        driver.createTopic("collisions", diaspora::Metadata{}, validator, selector, serializer);
        // END CREATE TOPIC

        // START ADD PARTITION
        auto& diaspora_driver = driver.as<mofka::MofkaDriver>();
        // add an in-memory partition
        diaspora_driver.addMemoryPartition("collisions", 0);
        // add a default partition (all arguments specified)
        diaspora_driver.addDefaultPartition(
                "collisions", 0,
                "my_metadata_provider@local",
                "my_data_provider@local",
                {}, "__primary__");
        // add a default partition (discover providers automatically)
        diaspora_driver.addDefaultPartition("collisions", 0);
        // END ADD PARTITION

        // START ADD PROVIDERS
        // add a metadata provider
        auto metadata_config = diaspora::Metadata{R"({
            "database": {
                "type": "log",
                "config": {
                    "path": "/tmp/diaspora-log",
                    "create_if_missing": true
                }
            }
        })"};
        auto metadata_provider = diaspora_driver.addDefaultMetadataProvider(0, metadata_config);
        // add a data provider
        auto data_config = diaspora::Metadata{R"({
            "target": {
                "type": "abtio",
                "config": {
                    "path": "/tmp/diaspora-data",
                    "create_if_missing": true
                }
            }
        })"};
        auto data_provider = diaspora_driver.addDefaultDataProvider(0, data_config);
        // create a partition that uses these providers
        diaspora_driver.addDefaultPartition("collisions", 0, metadata_provider, data_provider);
        // END ADD PROVIDERS
        {
        // START PRODUCER
        diaspora::TopicHandle topic = driver.openTopic("collisions");

        diaspora::ThreadPool thread_pool = driver.makeThreadPool(diaspora::ThreadCount{4});
        diaspora::BatchSize  batch_size = diaspora::BatchSize::Adaptive();
        diaspora::Ordering   ordering = diaspora::Ordering::Loose; // or Strict

        diaspora::Producer producer = topic.producer(
                "app1", thread_pool, batch_size, ordering);
        // END PRODUCER

        // START EVENT
        std::vector<char> segment1 = { 'a', 'b', 'c', 'd' };

        // expose 1 segment using its pointer and size
        diaspora::DataView data1{segment1.data(), segment1.size()};

        std::vector<char> segment2 = { 'e', 'f' };

        // expose 2 non-contiguous segments using diaspora::Data::Segment
        diaspora::DataView data2{{
            diaspora::DataView::Segment{segment1.data(), segment1.size()},
            diaspora::DataView::Segment{segment2.data(), segment2.size()}
        }};
        diaspora::Metadata metadata1{R"({"energy": 42})"};

        using json = nlohmann::json;
        auto md = json::object();
        md["energy"] = 42;
        diaspora::Metadata metadata2{md};
        // END EVENT

        // START PRODUCE EVENT
        diaspora::Future<diaspora::EventID> future = producer.push(metadata1, data1);
        future.completed(); // returns true if the future has completed

        producer.push(metadata2, data2);

        diaspora::EventID event_id_1 = future.wait();

        producer.flush();
        // END PRODUCE EVENT
        }

        {
        // START CONSUMER
        diaspora::TopicHandle topic = driver.openTopic("collisions");
        diaspora::BatchSize  batch_size = diaspora::BatchSize::Adaptive();
        diaspora::ThreadPool thread_pool = driver.defaultThreadPool();
        diaspora::DataSelector data_selector =
            [](const diaspora::Metadata& md, const diaspora::DataDescriptor& dd) -> diaspora::DataDescriptor {
                if(md.json()["energy"] > 20) {
                    return dd;
                } else {
                    return diaspora::DataDescriptor{};
                }
            };
        diaspora::DataAllocator data_allocator =
            [](const diaspora::Metadata& md, const diaspora::DataDescriptor& dd) -> diaspora::DataView {
                (void)md;
                char* ptr = new char[dd.size()];
                return diaspora::DataView{ptr, dd.size()};
            };
        diaspora::Consumer consumer = topic.consumer(
            "app2", thread_pool, batch_size, data_selector, data_allocator);
        // END CONSUMER
        // START CONSUME EVENTS
        diaspora::Future<diaspora::Event> future = consumer.pull();
        future.completed(); // returns true if the future has completed

        diaspora::Event event        = future.wait();
        diaspora::DataView data      = event.data();
        diaspora::Metadata metadata  = event.metadata();
        diaspora::EventID event_id   = event.id();

        event.acknowledge();

        delete[] static_cast<char*>(data.segments()[0].ptr);
        // END CONSUME EVENTS
        }

    } catch(const diaspora::Exception& ex) {
        std::cerr << ex.what() << std::endl;
    }

    return 0;
}

File: docs/_code/energy_topic.py
import sys
import pymargo.core
from pymargo.core import Engine
from mochi.mofka.client import MofkaDriver


def main(group_file: str):
    driver = MofkaDriver(group_file, use_progress_thread=True)

    # START CREATE TOPIC
    from mochi.mofka.client import Validator, PartitionSelector, Serializer

    validator = Validator.from_metadata(
        "energy_validator:libenergy_validator.so", {"energy_max": 100})
    selector = PartitionSelector.from_metadata(
        "energy_partition_selector:libenergy_partition_selector.so", {"energy_max": 100})
    serializer = Serializer.from_metadata(
        "energy_serializer:libenergy_serializer.so", {"energy_max": 100})

    driver.create_topic(
        topic_name="collisions",
        validator=validator,
        selector=selector,
        serializer=serializer)
    # END CREATE TOPIC

    # START ADD PARTITION
    # add an in-memory partition
    driver.add_memory_partition(
        topic_name="collisions",
        server_rank=0)
    # add a default partition (all arguments specified)
    driver.add_default_partition(
        topic_name="collisions",
        server_rank=0,
        metadata_provider="my_metadata_provider@local",
        data_provider="my_data_provider@local",
        partition_config={},
        pool_name="__primary__")
    # add a default partition (discover providers automatically)
    driver.add_default_partition(
        topic_name="collisions",
        server_rank=0)
    # END ADD PARTITION

    # START ADD PROVIDERS
    metadata_provider = driver.add_metadata_provider(
        server_rank=0,
        database_type="log",
        database_config={
            "path": "/tmp/mofka-log",
            "create_if_missing": True
        })
    data_provider = driver.add_data_provider(
        server_rank=0,
        target_type="abtio",
        target_config={
            "path": "/tmp/mofka-data",
            "create_if_missing": True
        })
    driver.add_default_partition(
        topic_name="collisions",
        server_rank=0,
        metadata_provider=metadata_provider,
        data_provider=data_provider)
    # END ADD PROVIDERS

    # START PRODUCER
    from mochi.mofka.client import ThreadPool, AdaptiveBatchSize, Ordering

    topic = driver.open_topic("collisions")
    thread_pool = ThreadPool(4)
    batch_size = AdaptiveBatchSize # or an integer > 0
    ordering = Ordering.Strict # or Ordering.Loose

    producer = topic.producer(
        name="app1",
        batch_size=batch_size,
        thread_pool=thread_pool,
        ordering=ordering)
    # END PRODUCER

    # START EVENT
    # data can be any object adhering to the buffer protocol
    data1 = b"abcd"
    data2 = bytearray("efgh", encoding="ascii")
    data3 = memoryview(data1)
    # or a list of such objects
    data4 = [data1, data2, data3]

    metadata1 = """{"energy": 42}""" # use a string
    metadata2 = {"energy": 42} # or use a dictionary
    # END EVENT

    # START PRODUCE EVENT
    future = producer.push(metadata=metadata1, data=data1)
    future.completed # returns True if the future has completed
    event_id = future.wait()

    producer.flush()
    # END PRODUCE EVENT

    # START CONSUMER
    from mochi.mofka.client import ThreadPool, AdaptiveBatchSize, DataDescriptor

    batch_size = AdaptiveBatchSize
    thread_pool = ThreadPool(0)

    def data_selector(metadata, descriptor):
        if metadata["energy"] > 20:
            return descriptor
        else:
            return None

    def data_broker(metadata, descriptor):
        # note that we return a *list* of objects satisfying the buffer protocol
        return [ bytearray(descriptor.size) ]

    consumer = topic.consumer(
        name="app2",
        thread_pool=thread_pool,
        batch_size=batch_size,
        data_selector=data_selector,
        data_broker=data_broker)
    # END CONSUMER

    # START CONSUME EVENTS
    future = consumer.pull()
    future.completed # returns true if the future has completed

    event    = future.wait()
    data     = event.data
    metadata = event.metadata
    event_id = event.event_id

    event.acknowledge()
    # END CONSUME EVENTS

if __name__ == "__main__":

    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <groupfile>")
    group_file = sys.argv[1]
    main(group_file)
